packages/react/presence/src/Presence.tsx

```diff
@@ -100,7 +101,10 @@ function usePresence(present: boolean) {
         const currentAnimationName = getAnimationName(stylesRef.current);
         const isCurrentAnimation = currentAnimationName.includes(event.animationName);
         if (event.target === node && isCurrentAnimation) {
-          send('ANIMATION_END');
+          // With React 18 concurrency this update is applied
+          // a frame after the animation ends, creating a flash of visible content.
+          // By manually flushing we ensure they sync within a frame, removing the flash.
+          ReactDOM.flushSync(() => send('ANIMATION_END'));
```

Comment:
Interesting…
I'm no unclear as to when `flushSync` should be used or not. 

As far as I understand, we use `flushSync` to opt-out of automatic batching.

But what's not clear to me here, is that if it was automatically batching and that was the issue, why would it be delayed by a frame? Unless it is batching 2 state updates "later", and opting out results in applying the first one at the correct time for us?

In which case that would kinda make sense.

Reply:
> But what's not clear to me here, is that if it was automatically batching and that was the issue, why would it be delayed by a frame?

This isn't entirely clear to me either, my current thought is that React is pushing this to the next tick as part of the batching mechanism, I saw this eluded to in the [wg post](https://github.com/reactwg/react-18/discussions/21):

> React doesn’t render the result of the first setState synchronously—the render occurs during the next browser tick. So the render hasn’t happened yet

Which I think is why we're seeing the state change trail the native event, here's a reduced example:

https://codesandbox.io/s/elated-snow-x36ypm?file=/src/App.js

Reply:
This is what I'm confused about though with that reduced example.
Once you isolate just the bad case (the red box), there's only really 1 setState right? So where's the batching? There's no batching in this case, or are we saying that React delays it always regardless of whether there are more than 1 updates?

Reply:
Man yeah definitely seems to be the case, it's pretty clear here.
It does make me think we probably need to do it in other places too then, as we use things like resize observer too…
The worse part in all that really, the part that bothers me is that it seems like there's no way to tell where it should be applied until you see an issue really…

Reply:
I hear you, though I do think we’ll begin to see best practices emerge over time through guidance from the core team and community (as we did with state, FCs, hooks etc). I’d like to better understand the underlying mechanics better though, right now it’s not particularly intuitive when to reach for certain tools.

Something that crossed my mind is where we explicitly opt out of batching using `setTimeout` when programatically focusing, nothing has come up yet though so we're probably ok. We might need to think about how to test for 17 if we decide to make changes in any spicy areas.